### PR TITLE
kubeadm: Generate kube-proxy image name in its own pkg

### PR DESCRIPTION
**Pull request by kad**
**What this PR does / why we need it**: code cleanup. Control kube-proxy image parameter in DaemonSet template.

**Which issue this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close that issue when PR gets merged)*: fixes #

**Special notes for your reviewer**:
Minor cleanup. Attn @luxas 

**Release note**:
```release-note
NONE
```

### CODE REVIEW

**Review by luxas — COMMENTED** [review]
I'd much rather cut all deps on `./app/images` from the proxy package and instead use `{{ .ImageRepository }}/kube-proxy-{{ .Arch }}:{{ .Version }}` in the proxy manifest.

Can you do that please?

**Comment by kad** [comment]
@luxas see also #48808. IMHO it is a lot easier to have ImageRepository  / version processing in one place rather than adding same safety or registry selection logic in multiple places.

**Comment by kad** [comment]
@luxas  PR updated with your request, however, I'd prefer that #48808 merged first, then this one rebased (as this one uses one commit from another).

File: cmd/kubeadm/app/phases/addons/manifests.go [hunk 1]
```diff
@@ -68,7 +68,7 @@ spec:
     spec:
       containers:
       - name: kube-proxy
-        image: {{ .Image }}
+        image: {{ if .ImageOverride }}{{ .ImageOverride }}{{ else }}{{ .ImageRepository }}/kube-proxy-{{ .Arch }}:{{ .Version }}{{ end }}
```
Comment (luxas) [inline cmd/kubeadm/app/phases/addons/manifests.go]: No need to support .ImageOverride here.
This will get it's own, overridable `.Image` field in the config file as we go. It shouldn't be tied to the hyperkube image as it's an addon. This was an oversight from earlier.

File: cmd/kubeadm/app/images/images.go [hunk 2]
```diff
@@ -39,10 +40,10 @@ func GetCoreImage(image string, cfg *kubeadmapi.MasterConfiguration, overrideIma
...
-		KubeSchedulerImage:         fmt.Sprintf("%s/%s-%s:%s", repoPrefix, "kube-scheduler", runtime.GOARCH, cfg.KubernetesVersion),
-		KubeProxyImage:             fmt.Sprintf("%s/%s-%s:%s", repoPrefix, "kube-proxy", runtime.GOARCH, cfg.KubernetesVersion),
+		KubeEtcdImage:              fmt.Sprintf("%s/%s-%s:%s", repoPrefix, "etcd", runtime.GOARCH, kubeadmutil.KubernetesVersionToImageTag(etcdVersion)),
+		KubeAPIServerImage:         fmt.Sprintf("%s/%s-%s:%s", repoPrefix, "kube-apiserver", runtime.GOARCH, kubeadmutil.KubernetesVersionToImageTag(cfg.KubernetesVersion)),
+		KubeControllerManagerImage: fmt.Sprintf("%s/%s-%s:%s", repoPrefix, "kube-controller-manager", runtime.GOARCH, kubeadmutil.KubernetesVersionToImageTag(cfg.KubernetesVersion)),
+		KubeSchedulerImage:         fmt.Sprintf("%s/%s-%s:%s", repoPrefix, "kube-scheduler", runtime.GOARCH, kubeadmutil.KubernetesVersionToImageTag(cfg.KubernetesVersion)),
+		KubeProxyImage:             fmt.Sprintf("%s/%s-%s:%s", repoPrefix, "kube-proxy", runtime.GOARCH, kubeadmutil.KubernetesVersionToImageTag(cfg.KubernetesVersion)),
```
Comment (luxas) [inline cmd/kubeadm/app/images/images.go]: I guess you can safely remove this case and constant now?

File: cmd/kubeadm/app/images/images.go [hunk 3]
```diff
@@ -39,10 +40,10 @@ func GetCoreImage(image string, cfg *kubeadmapi.MasterConfiguration, overrideIma
 	}
 	repoPrefix := kubeadmapi.GlobalEnvParams.RepositoryPrefix
 	return map[string]string{
-		KubeEtcdImage:              fmt.Sprintf("%s/%s-%s:%s", repoPrefix, "etcd", runtime.GOARCH, etcdVersion),
-		KubeAPIServerImage:         fmt.Sprintf("%s/%s-%s:%s", repoPrefix, "kube-apiserver", runtime.GOARCH, cfg.KubernetesVersion),
-		KubeControllerManagerImage: fmt.Sprintf("%s/%s-%s:%s", repoPrefix, "kube-controller-manager", runtime.GOARCH, cfg.KubernetesVersion),
-		KubeSchedulerImage:         fmt.Sprintf("%s/%s-%s:%s", repoPrefix, "kube-scheduler", runtime.GOARCH, cfg.KubernetesVersion),
-		KubeProxyImage:             fmt.Sprintf("%s/%s-%s:%s", repoPrefix, "kube-proxy", runtime.GOARCH, cfg.KubernetesVersion),
+		KubeEtcdImage:              fmt.Sprintf("%s/%s-%s:%s", repoPrefix, "etcd", runtime.GOARCH, kubeadmutil.KubernetesVersionToImageTag(etcdVersion)),
+		KubeAPIServerImage:         fmt.Sprintf("%s/%s-%s:%s", repoPrefix, "kube-apiserver", runtime.GOARCH, kubeadmutil.KubernetesVersionToImageTag(cfg.KubernetesVersion)),
```
Comment (luxas) [inline cmd/kubeadm/app/images/images.go]: might make sense to create a variable like `k8sImageTag` for `kubeadmutil.KubernetesVersionToImageTag(cfg.KubernetesVersion)` instead of having it on every case.

File: cmd/kubeadm/app/util/version.go [hunk 4]
```diff
@@ -72,3 +72,12 @@ func KubernetesReleaseVersion(version string) (string, error) {
 	}
 	return "", fmt.Errorf("version %q doesn't match patterns for neither semantic version nor labels (stable, latest, ...)", version)
 }
+
+// KubernetesVersionToImageTag is helper function that replaces all
+// non-allowed symbols in tag strings with underscores.
+// Image tag can only contain lowercase and uppercase letters, digits,
+// underscores, periods and dashes.
+func KubernetesVersionToImageTag(version string) string {
```
Comment (luxas) [inline cmd/kubeadm/app/util/version.go]: Keep in mind that this version already matches `kubeReleaseRegex` above. So the only disallowed char allowed in semver but not in an image tag is `+` AFAIK.
Therefore, could you make this clearer that the case here is about replacing `+` with `_` for the CI case where images are actually pushed with `_` instead of `+` instead of seemingly randomly converting strange chars to underscores.

**Review by luxas — COMMENTED** [review]
SGTM to let the other PR merge first.

File: cmd/kubeadm/app/images/images.go (hunk 3 above)
Comment (kad) [inline cmd/kubeadm/app/images/images.go]: yes, that can be optimized.

File: cmd/kubeadm/app/images/images.go (hunk 2 above)
Comment (kad) [inline cmd/kubeadm/app/images/images.go]: yep, will do.

File: cmd/kubeadm/app/phases/addons/manifests.go (hunk 1 above)
Comment (kad) [inline cmd/kubeadm/app/phases/addons/manifests.go]: for now, I think it's better to be consistent with current behavior, same as with GetCoreImage()
Altering that functionality better to be in own PR with separate changelog.

File: cmd/kubeadm/app/util/version.go (hunk 4 above)
Comment (kad) [inline cmd/kubeadm/app/util/version.go]: That's the case for current usage in one place where we know input. but in case of being public function of utils package, we never know what will be input but output will be tag in acceptable form. I'll add comment about known usage and semver.

**Comment by kad** [comment]
Rebased and cleaned up more unused code as commit 2c4a7c0

**Comment by kad** [comment]
rebased. can someone add "ok to test" please ?

**Review by luxas — APPROVED** [review]
/lgtm
/ok-to-test
